src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -100,6 +82,45 @@ public User actAs(X x, User entity) throws AuthenticationException {
     return user;
   }
 
+  public boolean canActAs(X x, User agent, User entity) {
+    try {
+      Group group = (Group) groupDAO_.find(entity.getGroup());
+      
+      if ( group == null ) {
+        throw new AuthorizationException("Entity must exist within a group.");
```

Comment:
Why are these exceptional cases rather than just returning false for "no you can't act as this non existant group, or no you can't act as a group that is disabled?"

Reply:
If we do remove the exceptions, we can remove them from the Interface declaration as well.

It seems easier to code against if its just a true/false value.  I think the key question is, is the client ever going to need to behave differently for cases where the group is disabled/doesn't exist as opposed to just not being a member of the appropriate group?  Most of the time I expect the client not to care, but I can imagine in a debugging scenario a more detailed error message would be useful.

Perhaps that's handled better by an error on the actual request though, if I try to make a request acting as user XYZ, it would be helpful to know that the request failed because there is no group XYZ, or group XYZ is disabled, meaning i likely have a configuration error or I picked the wrong group, as opposed to not being a member of group XYZ.

Upon further consideration, I'm not even sure that we should return an false if the group is disabled.  It seems reasonable to me to have canActAs return true, if I'm a member of the appropriate group, but to then have my requests fail later because the group is disabled.  If we ever add more states to whether a group is allowed to do perform actions we're not going to want to come back and update this method as well, I think this method should really just be telling me (You should be able to make requests on behalf of user XYZ, whether or not those will succeed due to the group being disabled or over quota is a different question)

Reply:
> Why are these exceptional cases rather than just returning false for "no you can't act as this non existant group, or no you can't act as a group that is disabled?"

Just to clarify, you're not acting as a group, you're acting as another User (or in Ablii's case a Business, which extends the User model).

One of the requirements of Ablii was that users (people) should be able to be part of multiple businesses. The way we implemented that was allowing `User`s to "act as" other `User`s. 

Another requirement was that users will need different permissions for different businesses. We implemented that by storing a group id on each junction between a user and a business. That group becomes the "effective" user's group while acting as a business. The junction group is the effective group, not the `User`'s group and not the `Business`'s group.

We considered it exceptional to have an invalid group on a junction when trying to act as another User because that's the group that's the "effective" group, the group that's going to be checked by the auth service. So if the group referenced on the junction doesn't exist, then you're effectively not going to have a group.

I think your last paragraph is reasonable though, I just wanted to clarify the requirements we were working under and how it currently works.